consensus/src/alerts/handler.rs

```diff
@@ -0,0 +1,752 @@
+                return None;
+            }
+        };
+        let contents = alert.as_signable();
+        if let Some(forker) = self.who_is_forking(&contents.proof) {
+            if self.known_rmcs.contains_key(&(contents.sender, forker)) {
+                debug!(target: "AlephBFT-alerter","{:?} We already know about an alert by {:?} about {:?}.", self.index(), alert.as_signable().sender, forker);
+                self.known_alerts.insert(contents.hash(), alert);
+                return None;
+            }
+            let propagate_alert = if self.is_forker(forker) {
+                None
+            } else {
+                // We learn about this forker for the first time, need to send our own alert
+                self.on_new_forker_detected(forker, contents.proof.clone());
+                Some(ForkingNotification::Forker(contents.proof.clone()))
+            };
+            let hash_for_rmc = self.rmc_alert(forker, alert);
+            Some((propagate_alert, hash_for_rmc))
+        } else {
+            warn!(target: "AlephBFT-alerter","{:?} We have received an incorrect forking proof from {:?}.", self.index(), alert.as_signable().sender);
+            None
+        }
+    }
+
+    /// `on_message()` may return an `AlerterResponse` which should be propagated
+    pub fn on_message(
+        &mut self,
+        message: AlertMessage<H, D, MK::Signature, MK::PartialMultisignature>,
+    ) -> Option<AlerterResponse<H, D, MK::Signature, MK::PartialMultisignature>> {
+        use AlertMessage::*;
+        match message {
+            ForkAlert(alert) => {
+                trace!(target: "AlephBFT-alerter", "{:?} Fork alert received {:?}.", self.index(), alert);
+                self.on_network_alert(alert)
+                    .map(|(n, h)| AlerterResponse::ForkResponse(n, h))
+            }
+            RmcMessage(sender, message) => {
+                let hash = message.hash();
+                if let Some(alert) = self.known_alerts.get(hash) {
+                    let alert_id = (alert.as_signable().sender, alert.as_signable().forker());
+                    if self.known_rmcs.get(&alert_id) == Some(hash) || message.is_complete() {
+                        Some(AlerterResponse::RmcMessage(message))
+                    } else {
+                        None
+                    }
+                } else {
+                    Some(AlerterResponse::AlertRequest(
+                        *hash,
+                        Recipient::Node(sender),
+                    ))
+                }
+            }
+            AlertRequest(node, hash) => match self.known_alerts.get(&hash) {
+                Some(alert) => Some(AlerterResponse::ForkAlert(
+                    alert.clone().into_unchecked(),
+                    Recipient::Node(node),
+                )),
+                None => {
+                    debug!(target: "AlephBFT-alerter", "{:?} Received request for unknown alert.", self.index());
```

Comment:
Hm, less sure whether this should also be an error, but perhaps? Please look into this wile changing the other things.

Reply:
we will probably need a `Result<Option<AlerterResponse>, Error>`. The first `None` I treat as no error so we would return `Ok(None)` and the last `None` would be some kind of error.